text/0000-delegation-of-implementation.md

```diff
@@ -0,0 +1,235 @@
+- Feature Name: delegation_of_implementation
+- Start Date: 2015-12-12
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+[summary]: #summary
+
+Provide a syntactic sugar to automatically implement a given trait `Tr` using a pre-existing type implementing `Tr`. The purpose is to improve code reuse in rust without damaging the orthogonality of already existing concepts or adding new ones.
+
+# Motivation
+[motivation]: #motivation
+
+Let's consider some existing pieces of code:
+```rust
+// from rust/src/test/run-pass/dropck_legal_cycles.rs
+impl<'a> Hash for H<'a> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state)
+    }
+}
+```
+```rust
+// from servo/components/devtools/actors/timeline.rs
+impl Encodable for HighResolutionStamp {
+    fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
+        self.0.encode(s)
+    }
+}
+```
+We can see a recurring pattern where the implementation of a method only consists in applying the same method to a subfield or more generally to an expression containing `self`. Those are examples of the well known [composition pattern][object_composition]. It has a lot of advantages but unfortunately it also implies writing explicit boilerplate code again and again. In a classical oop language we could also have opted for inheritance for similar cases. Inheritance comes with its own bunch of problems and limitations but at least it allows a straightforward form of code reuse: any subclass implicitly imports the public methods of its superclass(es).
+
+Rust has no inheritance (yet) and as a result composition is an even more interesting pattern for factoring code than in other languages. In fact it is already used in many places. Some (approximate) figures:
+
+Project           | Occurences of "delegating methods" |
```

Comment:
Percentages would be much more useful than numbers here, I don't know what these values mean without a referent.